app/Mail/Services/UnsubscribeService.php

```diff
@@ -0,0 +1,283 @@
+            // This handles both explicit subscriptions (updates them) and implicit
+            // subscriptions (creates an explicit unsubscribe record).
+            Subscription::updateOrCreate(
+                [
+                    'user_id' => $user->id,
+                    'subject_type' => $data->subjectType,
+                    'subject_id' => $data->subjectId,
+                ],
+                [
+                    'state' => false, // set to false rather than deleting to prevent implicit re-subscription
+                ]
+            );
+
+            $descriptionData = $this->getGranularDescription($data->subjectType, $data->subjectId);
+        } elseif ($data instanceof CategoryUnsubscribeData) {
+            $currentPrefs = $user->websitePrefs;
+            $newPrefs = $currentPrefs & ~(1 << $data->preference);
+            $user->websitePrefs = $newPrefs;
+            $user->save();
+
+            $descriptionData = $this->getCategoryDescription($data->preference);
+        }
+
+        return [
+            'success' => true,
+            'descriptionKey' => $descriptionData['key'],
+            'descriptionParams' => $descriptionData['params'] ?? null,
+            'undoToken' => $undoToken,
+            'user' => $user,
+        ];
+    }
+
+    /**
+     * Generate an undo token that expires in 24 hours.
+     */
+    public function generateUndoToken(UnsubscribeData $data): string
+    {
+        $token = bin2hex(random_bytes(32));
+
+        Cache::put(
+            CacheKey::buildUnsubscribeUndoTokenCacheKey($token),
+            $data->toJson(),
+            now()->addHours(24)
+        );
+
+        return $token;
+    }
+
+    public function processUndo(string $token): array
+    {
+        $json = Cache::get(CacheKey::buildUnsubscribeUndoTokenCacheKey($token));
+
+        if (!$json) {
+            return ['success' => false, 'errorCode' => 'undo_expired'];
+        }
+
+        try {
+            $payload = json_decode($json, true);
+
+            $data = match ($payload['type']) {
+                'granular' => GranularUnsubscribeData::from($payload),
+                'category' => CategoryUnsubscribeData::from($payload),
+                default => null,
+            };
+
+            if (!$data) {
+                return ['success' => false, 'errorCode' => 'invalid_undo'];
+            }
+        } catch (Exception $e) {
+            return ['success' => false, 'errorCode' => 'invalid_undo_format'];
+        }
+
+        $user = User::find($data->userId);
+        if (!$user) {
+            return ['success' => false, 'errorCode' => 'user_not_found'];
+        }
+
+        if ($data instanceof GranularUnsubscribeData) {
+            // Remove the explicit unsubscribe record to restore the subscription.
+            // This will revert to implicit subscription if the user has commented,
+            // or no subscription if they haven't.
```

Comment:
What if they had an explicit subscription before?
1) Subscribe to game wall X
2) Get email for game wall X
3) Unsubscribe from game wall X
4) Undo

The user is no longer subscribed at this point.